Add set_unicode_input_mode_noeeprom, and update docs.

zunger-humu opened the pull request:
## Description
Make it clearer how to set the unicode mode from a keymap/keyboard without thrashing the EEPROM.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [X] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [X] Documentation


## Issues Fixed or Closed by this PR

* 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qmk/qmk_firmware/4937)
<!-- Reviewable:end -->

vomindoraan commented:
We've discussed doing this previously, and we wrote it off as unnecessary. What's the use case? Unless you're changing the input mode thousands of times, writing to EEPROM shouldn't be an issue.

vomindoraan commented on docs/feature_unicode.md:
````diff
@@ -97,13 +102,7 @@ You can switch the input mode at any time by using one of the following keycodes
-  set_unicode_input_mode(UC_LNX);
-}
-```
+You can also switch the input mode by calling `set_unicode_input_mode(x)` or `set_unicode_input_mode_noeeprom(x)` in your code, where _x_ is one of the above input mode constants (e.g. `UC_LNX`). Since the first of these writes to EEPROM, you shouldn't call it unless the value of the mode has actually changed. If you want to set the value at keyboard power-on time, you can examine the value alrea ...
````
I feel this needs more clarification and/or is misleading.

What do you mean by “unless the value of the mode has actually changed”? The values of the `UC_` constants are fixed and likely won't be changing for this exact reason, so there's no need to mention it in the docs, if that's what you meant.

Using `set_unicode_input_mode_noeeprom` for changing to your preferred input mode when the board starts is kind of missing the point. This is already done by `unicode_input_mode_init` — if you want the board to start with your preferred input mode, just add `#define UNICODE_SELECTED_MODES <your mode>` to your `config.h`. You don't need to set it yourself. (Granted, this should be made clearer in the docs.)

The intended use case for `set_unicode_input_mode` is to programmatically set the input mode, e.g. from a custom keycode or other function. For example, you want `MY_KEYCODE` to switch to `UC_LNX` if you're holding Shift and `UC_WINC` if you're holding Ctrl. The difference with `_noeeprom` is that the change isn't persisted, and the board will boot up with the previous value. This paragraph should focus on that instead of explaining how to set the mode on power up.

drashna commented:
Actually, this may not be needed, anymore.

On AVR, it looks like the "update" function that we use for EEPROM will actually check the existing value and will not write if it is the same value.  

ARM didn't do this, but I think that's already been fixed. 

So... there should be no harm in calling this on every matrix init.

zunger-humu commented:
In either case, it sounds like I should be updating the documentation to
reflect the "right" way to absolute-set the unicode mode. It sounds like
the two options are to call `set_unicode_input_mode` in matrix_init_user,
or to #define UNICODE_SELECTED_MODES MY_MODE. Konstantin, which would you
recommend? (I'll change this to a docs-only change)

On Wed, Jan 30, 2019 at 11:33 AM Drashna Jaelre <notifications@github.com>
wrote:

> Actually, this may not be needed, anymore.
>
> On AVR, it looks like the "update" function that we use for EEPROM will
> actually check the existing value and will not write if it is the same
> value.
>
> ARM didn't do this, but I think that's already been fixed.
>
> So... there should be no harm in calling this on every matrix init.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/4937#issuecomment-459078498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdpQl-gP3MIBrva25Phbn5J4_p-q3_2kks5vIfOngaJpZM4aSNM->
> .
>

vomindoraan commented:
I would recommend `UNICODE_SELECTED_MODES`. Two major reasons and one minor reason:

- It's declarative and easy to use. All the user needs to do is add one line to their `config.h`, and the rest is taken care of for them. Being a `config.h` setting, you also get other benefits, like overridability (e.g. you can set your most commonly used modes in your userspace to be shared across your boards, but then you can override it for specific boards by redefining the setting in the keymap's `config.h`).
- You get input mode toggling/cycling for free. For example, `#define UNICODE_SELECTED_MODES UC_LNX, UC_WINC` lets you toggle between those two modes with the `UC_MOD` keycode. The board also remembers which mode you used last and boots with it the next time you plug it in (this can be disabled with `#define UNICODE_CYCLE_PERSIST false`; in that case, the board will always boot to the first mode in the list).

And the minor reason:

- Initializing the input mode this way doesn't write to EEPROM, whereas setting it in `matrix_init_user` writes to EEPROM at least once every time the board boots.

Oh and by the way, thank you for offering to update the docs on this! I appreciate it, as I never got around to doing it myself 😄

zunger-humu commented:
OK, cool! I removed the `_noeeprom` function, updated the docs per our conversation, and also changed the implementation in my recent keymap to illustrate the proper behavior. PTAL.

drashna commented:
There are some merge conflicts here that can't be resolved here.  

You may need to rebase your branch to fix this.

zunger-humu commented:
Rebased.

zunger-humu commented:
@vomindoraan @drashna One of you needs to squash and merge this, right?

drashna commented on docs/feature_unicode.md:
````diff
@@ -97,12 +101,10 @@ You can switch the input mode at any time by using one of the following keycodes
+To explicitly change the mode and update the EEPROM in code, call `set_unicode_input_mode(x)` with one of the input modes. To lock the mode to a fixed value in your keymapor keyboard, simply enable input method cycling with only one mode available, e.g. in a `config.h` add
 
-```c
-void eeconfig_init_user(void) {
````
This shouldn't be removed, actually.  

It sets the default mode if the eeprom is reset.  
And when you reset the eeprom, it doesn't re-run the init code.  So .... this is probably the better way.  ESPECIALLY, if you want to use multiple methods (but want a default, still). 

But these are not mutually exclusive. options.

drashna commented on docs/feature_unicode.md:
```diff
@@ -83,9 +83,13 @@ The following input modes are available:
 ### Switching Input Modes
 
-There are two ways to set the input mode for Unicode: by keycode or by function. Keep in mind that both methods write to persistent storage (EEPROM), and are loaded each time the keyboard starts. So once you've set it the first time, you don't need to set it again unless you want to change it, or you've reset the EEPROM settings.
+The Unicode input mode is stored in two places: in persistent storage (EEPROM) and in memory. The system normally reads the value from EEPROM at init time, and relies on the in-memory value thereafter.
```
I'll have to take a look at this again, but .... I'm not happy with the wording here.  By itself ... it's okay. But when taken as a whole, it doesn't flow correctly.

drashna commented:
There are some merge conflicts here, now.
